x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/WaitForSnapshotInProgressStep.java

```diff
@@ -0,0 +1,127 @@
+import org.elasticsearch.snapshots.SnapshotsService;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Locale;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+/**
+ * After we triggered a snapshot we'll wait until the snapshot is not executing anymore (ie. the SnapshotInProgress meta data doesn't
+ * contain an entry for the snapshot we've triggered)
+ */
+public class WaitForSnapshotInProgressStep extends ClusterStateWaitStep {
+
+    public static final String NAME = "wait-for-snapshot";
+    private static final Logger logger = LogManager.getLogger(WaitForSnapshotInProgressStep.class);
+
+    WaitForSnapshotInProgressStep(StepKey key, StepKey nextStepKey) {
+        super(key, nextStepKey);
+    }
+
+    @Override
+    public boolean isRetryable() {
+        return true;
+    }
+
+    @Override
+    public Result isConditionMet(Index index, ClusterState clusterState) {
+        IndexMetaData indexMetaData = clusterState.metaData().index(index);
+        String indexName = index.getName();
+
+        if (indexMetaData == null) {
+            String errorMessage = String.format(Locale.ROOT, "[%s] lifecycle action for index [%s] executed but index no longer exists",
+                getKey().getAction(), indexName);
+            // Index must have been since deleted
+            logger.debug(errorMessage);
+            return new Result(false, new Info(errorMessage));
+        }
+
+        String policyName = indexMetaData.getSettings().get(LifecycleSettings.LIFECYCLE_NAME);
+        LifecycleExecutionState executionState = LifecycleExecutionState.fromIndexMetadata(indexMetaData);
+        final String snapshotRepository = executionState.getSnapshotRepository();
+        if (Strings.hasText(snapshotRepository) == false) {
+            String errorMessage = String.format(Locale.ROOT, "snapshot repository is not present for policy [%s] and index [%s]",
+                policyName, indexName);
+            logger.debug(errorMessage);
+            return new Result(false, new Info(errorMessage));
+        }
+        final String snapshotName = executionState.getSnapshotName();
+        if (Strings.hasText(snapshotName) == false) {
+            String errorMessage = String.format(Locale.ROOT, "snapshot name was not generated for policy [%s] and index [%s]",
+                policyName, indexName);
+            logger.debug(errorMessage);
+            return new Result(false, new Info(errorMessage));
+        }
+        final SnapshotsInProgress snapshotsInProgress = clusterState.custom(SnapshotsInProgress.TYPE);
+        // we got here before the cluster state was updated with the (or any) snapshot being started, so we'll wait
+        if (snapshotsInProgress == null) {
+            return new Result(false, new Info(String.format(Locale.ROOT,
+                "snapshot [%s] generated by policy [%s] for index [%s] is still in progress", snapshotName, policyName, indexName)));
```

Comment:
I actually think we need to return `true` here, it's possible that the snapshot was taken and completed (prior to the step being updated), and then the cluster was restarted, so the `SnapshotsInProgress` is null and won't ever exist because the snapshot was successful, best to err on the side of moving forward and doing a check than be wedged I think

Reply:
Deleted the WaitForSnapshotInProgressStep altogether as we'll design a step that waits for a generation change in the repo metadata in a subsequent PR.